File: src/SoulCharger.ts
import Mod from "./Mod";
import { clamp } from "./util";

const SOUL_SPEED = 8;

export default class SoulCharger {
  private soulEffects = new Set<int>();

  constructor(
    private mod: Mod,
    private owner: EntityPlayer,
    private maxCharge: int,
  ) {}

  Load(): void {
    this.AddCallbacks();
  }

  Unload(): void {
    this.RemoveCallbacks();
  }

  private AddCallbacks(): void {
    Mod.AddCallback(ModCallbacks.MC_POST_NPC_DEATH, this.PostNpcDeath);

    Mod.AddCallback(
      ModCallbacks.MC_POST_EFFECT_UPDATE,
      this.PostEffectUpdate,
      EffectVariant.ENEMY_SOUL,
    );
  }

  private RemoveCallbacks(): void {
    Mod.RemoveCallback(ModCallbacks.MC_POST_NPC_DEATH, this.PostNpcDeath);

    Mod.RemoveCallback(
      ModCallbacks.MC_POST_EFFECT_UPDATE,
      this.PostEffectUpdate,
    );
  }

  PostNpcDeath = (enemy: EntityNPC): void => {
    if (!enemy.IsActiveEnemy(true)) return;

    let effect = Isaac.Spawn(
      EntityType.ENTITY_EFFECT,
      EffectVariant.ENEMY_SOUL,
      0,
      enemy.Position,
      Vector.Zero,
      this.owner,
    ).ToEffect()!;

    this.soulEffects.add(effect.Index);
  };

  PostEffectUpdate = (effect: EntityEffect): void => {
    if (!this.soulEffects.has(effect.Index)) return;

    if (!effect.Exists()) {
      this.soulEffects.delete(effect.Index);
      return;
    }

    let direction = this.owner.Position.sub(effect.Position).Normalized();

    effect.Velocity = direction.mul(SOUL_SPEED);
    if (effect.Position.DistanceSquared(this.owner.Position) < 400) {
      effect.Remove();
      Isaac.Spawn(
        EntityType.ENTITY_EFFECT,
        EffectVariant.COIN_PARTICLE,
        0,
        this.owner.Position,
        Vector.Zero,
        effect,
      );
      this.addCharge(1);
    }
  };

  private getCharge(): int {
    return this.owner.GetActiveCharge(ActiveSlot.SLOT_PRIMARY);
  }

  private addCharge(charge: number): void {
    let newCharge = clamp(this.getCharge() + charge, 0, this.maxCharge);
    this.owner.SetActiveCharge(newCharge, ActiveSlot.SLOT_PRIMARY);
  }
}

File: mod/code/SoulCharger.lua
local util = include("code.util")

local SOUL_SPEED = 8

local SoulCharger = {}
SoulCharger.__index = SoulCharger

function SoulCharger.new(mod, owner, maxCharge)
  local self = {
    mod = mod,
    owner = owner,
    maxCharge = maxCharge,
    activeSlot = ActiveSlot.SLOT_PRIMARY,

    soulEffects = {},
  }

  function self.PostNpcDeath(_, enemy)
    if not enemy:IsActiveEnemy(true) then return end

    local effect = Isaac.Spawn(
      EntityType.ENTITY_EFFECT,
      EffectVariant.ENEMY_SOUL,
      0,
      enemy.Position,
      Vector.Zero,
      this.owner
    ):ToEffect()

    self.soulEffects[effect.Index] = true
  end

  function self.PostEffectUpdate(_, effect)
    if not self.soulEffects[effect.Index] then return end

    if not effect:Exists() then
      self.soulEffects[effect.Index] = nil
      return
    end

    local direction = (self.owner.Position - effect.Position):Normalized()

    effect.Velocity = direction * SOUL_SPEED
    if effect.Position:DistanceSquared(self.owner.Position) < 400 then
      effect:Remove()
      Isaac.Spawn(
        EntityType.ENTITY_EFFECT,
        EffectVariant.COIN_PARTICLE,
        0,
        this.owner.Position,
        Vector.Zero,
        effect
      )
      self:addCharge(1)
    end
  end

  setmetatable(self, SoulCharger)
  return self
end

function SoulCharger:Load()
  self:addCallbacks()
end

function SoulCharger:Unload()
  self:removeCallbacks()
end

function SoulCharger:addCallbacks()
  self.mod:AddCallback(ModCallbacks.MC_POST_NPC_DEATH, self.PostNpcDeath)

  self.mod:AddCallback(
      ModCallbacks.MC_POST_EFFECT_UPDATE,
      self.PostEffectUpdate,
      EffectVariant.ENEMY_SOUL
    )
end

function SoulCharger:removeCallbacks()
  self.mod:RemoveCallback(ModCallbacks.MC_POST_NPC_DEATH, self.PostNpcDeath)

  self.mod:RemoveCallback(
    ModCallbacks.MC_POST_EFFECT_UPDATE,
    self.PostEffectUpdate
  )
end

function SoulCharger:getCharge()
  return self.owner:GetActiveCharge(self.activeSlot)
end

function SoulCharger:addCharge(charge)
  local newCharge = util.clamp(self:getCharge() + charge, 0, self.maxCharge)
  self.owner:SetActiveCharge(newCharge, self.activeSlot)
end

return SoulCharger